kernel/fs.c

```diff
@@ -402,41 +402,122 @@ bmap(struct inode *ip, uint bn)
   panic("bmap: out of range");
 }
 
-// Truncate inode (discard contents).
-// Only called when the inode has no links
-// to it (no directory entries referring to it)
-// and has no in-memory reference to it (is
-// not an open file or current directory).
 static void
-itrunc(struct inode *ip)
-{
-  int i, j;
+new_space_inode(struct inode *ip, uint new){
+  uint sblk, eblk;
+  uint soff, eoff;
+  uint cur, newblk;
+  struct buf *bp;
+
+  if ( ip->size > new )
+    panic("new space inode");
+
+  if ( ip->size == new )
+    return;
+
+  sblk = ip->size / BSIZE; 
+  soff = ip->size % BSIZE;
+  eblk = new / BSIZE; 
+  eoff = new % BSIZE;
+
+  if ( sblk == eblk )
+    goto out;
+
+  if ( eoff > 0 )
+    ++eblk;
+
+  if ( soff > 0 ) {
+
+    bp = bread(ip->dev, bmap(ip, sblk));
+    memset(bp->data + soff, 0, BSIZE-soff);
+    log_write(bp);
+    brelse(bp);
+    ++sblk;
+  }
+
+  for( cur = sblk; eblk >= cur; ++cur) {
+
+    newblk = bmap(ip, cur);
+    if ( newblk == 0 )
+      panic("new_space_inode");
+  }
+
+out:
+  ip->size = new;
+  ip->mtime = cmosepochtime();
+  iupdate(ip);
+}
+
+static void
+free_space_inode(struct inode *ip, uint new){
+  uint sblk, eblk;
+  uint soff, eoff;
+  uint cur, curidx;
   struct buf *bp;
   uint *a;
 
-  for(i = 0; i < NDIRECT; i++){
-    if(ip->addrs[i]){
-      bfree(ip->dev, ip->addrs[i]);
-      ip->addrs[i] = 0;
+  if ( new > ip->size )
+    panic("free space inode");
+
+  sblk = ip->size / BSIZE; 
+  soff = ip->size % BSIZE;
+  eblk = new / BSIZE; 
+  eoff = new % BSIZE;
+  if ( eoff > 0 )
+    ++eblk;
+
+  if ( soff > 0 ) {
+
+    bp = bread(ip->dev, bmap(ip, sblk));
+    memset(bp->data + soff, 0, BSIZE-soff);
+    log_write(bp);
+    brelse(bp);
+    ++sblk;
+  }
+  for( cur = sblk; eblk >= cur; ++cur) {
```

Comment:
Does this loop works as intended?

This function is called only when `new >= ip->size`, so `sblk >= eblk`.
So the loop never happens and no block released.

Reply:
I forgot erasing this.
Thanks,

Reply:
I mean it is needed to swap start and end values each other:

```c
  for( cur = eblk; sblk >= cur; ++cur) {
```